rust/src/ffi/exported.rs

```diff
@@ -0,0 +1,244 @@
+// Copyright (c) 2025 Sonic Operations Ltd
+//
+// Use of this software is governed by the Business Source License included
+// in the LICENSE file and at soniclabs.com/bsl11.
+//
+// Change Date: 2028-4-16
+//
+// On the date above, in accordance with the Business Source License, use of
+// this software will be governed by the GNU Lesser General Public License v3.
+#![allow(unused_variables)]
+
+use std::ffi::{c_char, c_int, c_void};
+
+use crate::ffi::bindings;
+
+/// Opens a new state object based on the provided implementation maintaining
+/// its data in the given directory. If the directory does not exist, it is
+/// created. If it is empty, a new, empty state is initialized. If it contains
+/// state information, the information is loaded.
+///
+/// The function returns an opaque pointer to a state object that can be used
+/// with the remaining functions in this file. Ownership is transferred to the
+/// caller, which is required for releasing it eventually using Carmen_Rust_ReleaseState.
+/// If for some reason the creation of the state instance failed, a nullptr is
+/// returned.
+#[unsafe(no_mangle)]
+unsafe extern "C" fn Carmen_Rust_OpenState(
+    schema: u8,
+    state: bindings::StateImpl,
+    archive: bindings::ArchiveImpl,
+    directory: *const c_char,
+    length: c_int,
+) -> *mut c_void {
+    unimplemented!()
+}
+
+/// Flushes all committed state information to disk to guarantee permanent
+/// storage. All internally cached modifications are synced to disk.
+#[unsafe(no_mangle)]
+unsafe extern "C" fn Carmen_Rust_Flush(state: *mut c_void) {
+    unimplemented!()
+}
+
+/// Closes this state, releasing all IO handles and locks on external resources.
+#[unsafe(no_mangle)]
+unsafe extern "C" fn Carmen_Rust_Close(state: *mut c_void) {
+    unimplemented!()
+}
+
+/// Releases a state object, thereby causing its destruction. After releasing it,
+/// no more operations may be applied on it.
+#[unsafe(no_mangle)]
+unsafe extern "C" fn Carmen_Rust_ReleaseState(state: *mut c_void) {
+    unimplemented!()
+}
+
+/// Creates a state snapshot reflecting the state at the given block height. The
+/// resulting state must be released and must not outlive the life time of the
+/// provided state.
+#[unsafe(no_mangle)]
+unsafe extern "C" fn Carmen_Rust_GetArchiveState(state: *mut c_void, block: u64) -> *mut c_void {
+    unimplemented!()
+}
+
+/// Gets the current state of the given account.
+#[unsafe(no_mangle)]
+unsafe extern "C" fn Carmen_Rust_GetAccountState(
+    state: *mut c_void,
+    addr: *mut c_void,
+    out_state: *mut c_void,
```

Comment:
Rust question: would [type aliases](https://doc.rust-lang.org/reference/items/type-aliases.html) for the various object types work here? It would help readability a bit, but would it be Rusty?

The type for the address may, for instance, be defined as a pointer to a 20-byte array. Balances are pointers to 16-byte arrays, and nonces to an 8-byte array. This would be more precise then the generic void pointer.

I understand that the typing is not enforced on the C interface, but it may help adding some type checks to implementations.